packages/eslint-plugin/src/rules/consistent-indexed-object-style.ts

```diff
@@ -0,0 +1,118 @@
+import { createRule } from '../util';
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+export default createRule({
+  name: 'consistent-indexed-object-style',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Enforce or disallow the use of the record type',
+      category: 'Stylistic Issues',
+      // too opinionated to be recommended
+      recommended: false,
+    },
+    messages: {
+      preferRecord: 'A record is preferred over an index signature',
+      preferIndexSignature: 'An index signature is preferred over a record.',
+    },
+    fixable: 'code',
+    schema: [
+      {
+        enum: ['always', 'never'],
+      },
+    ],
+  },
+  defaultOptions: ['always'],
+  create(context) {
+    const sourceCode = context.getSourceCode();
+
+    if (context.options[0] === 'never') {
+      return {
+        TSTypeReference(node): void {
+          const typeName = node.typeName as TSESTree.Identifier;
```

Comment:
this is not a valid cast.

Example:
```ts
namespace A { export type B = 1 }
type T = A.B;
```

```suggestion
          const typeName = node.typeName;
          if (typeName.type !== AST_NODE_TYPES.Identifier) {
            return;
          }
```

Reply:
Good catch.